sensors: test: Add attribute support to sensor emulator backend

yperess opened the pull request:
* Fix BMI160 to align with data sheet
* Add BMI160 emulator backend so the data collection tests can be done by the generic test
* Add attribute 'get' support to BMI160 and emulator backend
* Write custom tests for bus failures on the BMI160

Commit:
sensor: testing: Update sensor emul backend

Update the backend for sensor emulators to include a function for
setting the offset as well as a function to query an attribute's
metadata such as bounds and increment size. Additionally, add
backend support for setting the _xyz channel values.

Make the appropriate test changes to accomodate.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
bmi160: Make changes to align driver to datasheet

The logic in the driver was not aligned to the datasheet. Also,
temperature reading was not being done in fetch, but in channel_get.
There was also some extra conversions from SI->register->SI when
setting the range, this was causing the register value calculation to
produce an incorrect scale in some cases.

Tests were added to cover these cases.

Signed-off-by: Yuval Peress <peress@google.com>

MaureenHelm commented on drivers/i2c/i2c_emul.c:
```diff
@@ -90,12 +90,15 @@ static int i2c_emul_transfer(const struct device *dev, struct i2c_msg *msgs, uin
-	if (ret) {
-		return ret;
+	if (emul->mock_api != NULL && emul->mock_api->transfer != NULL) {
+		ret = emul->mock_api->transfer(emul->target, msgs, num_msgs, addr);
+		printk("mock_api returned %d\n", ret);
+		if (ret != -ENOSYS) {
+			return ret;
+		}
 	}
 
-	return 0;
+	return api->transfer(emul->target, msgs, num_msgs, addr);
```
This change looks unrelated to attributes?

yperess replied:
Sorry, it's a part of the tests we're using in a related change downstream. I moved one such test upstream to use this as an example (more to follow).

MaureenHelm commented on drivers/i2c/i2c_emul.c:
```diff
@@ -125,7 +128,7 @@ int i2c_emul_register(const struct device *dev, struct i2c_emul *emul)
 	sys_slist_append(&data->emuls, &emul->node);
 
-	LOG_INF("Register emulator '%s' at I2C addr %02x\n", name, emul->addr);
+	LOG_INF("Register emulator '%s' at I2C addr %02x", name, emul->addr);
```
This change is definitely not related to attributes

MaureenHelm commented on drivers/sensor/akm09918c/akm09918c_emul.c:
```diff
@@ -162,8 +162,9 @@ static int akm09918c_emul_backend_set_channel(const struct emul *target, enum se
-			     1000000 / ((int64_t)INT32_MAX + 1);
+	int32_t microgauss =
+		(shift < 0 ? ((int64_t)*value >> -shift) : ((int64_t)*value << shift)) * 1000000 /
+		((int64_t)INT32_MAX + 1);
```
More unrelated clang-format changes

yperess replied:
Not unrelated `value` is now a pointer. Adding `*` required a format change due to line length.

MaureenHelm commented on include/zephyr/drivers/i2c_emul.h:
```diff
@@ -42,6 +43,12 @@ struct i2c_emul {
+	 * A mock API that if not NULL will take precedence over the actual API. If set, a return
+	 * value of -ENOSYS will revert back to the default api.
+	 */
+	struct i2c_emul_api *mock_api;
```
I2C changes should be in their own commit

MaureenHelm commented on drivers/sensor/bmi160/bmi160.c:
```diff
@@ -709,6 +724,128 @@ static int bmi160_attr_set(const struct device *dev, enum sensor_channel chan,
 	return 0;
 }
 
+static int bmi160_attr_get(const struct device *dev, enum sensor_channel chan,
```
New feature isn't related to the commit message `bmi160: Make changes to align driver to datasheet`

Commit:
bmi160: Implement attr_get

Add support for getting the following attribute values:
- SENSOR_ATTR_OFFSET
- SENSOR_ATTR_SAMPLING_FREQUENCY
- SENSOR_ATTR_FULL_SCALE

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
bus: emul: Update i2c/spi emulators with mock transport

Adding a hook for tests to inject a mock transport and migrating the
accel test to test bmi160 specific things. The old version of the test
which checks for read values is now covered by the generic test in
the sensor build_all target.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
test: bmi160: rename test directory

Rename the test directory to more accuratly reflect what's being
tested.

Signed-off-by: Yuval Peress <peress@google.com>

alevkoy commented on drivers/sensor/bmi160/emul_bmi160.h:
```diff
@@ -0,0 +1,61 @@
+extern "C" {
+#endif
+
+__maybe_unused static bool emul_bmi160_i2c_is_touching_reg(struct i2c_msg *msgs, int num_msgs,
```
Why are these defined in the header file?

yperess replied:
They're currently used in the bmi160 test suite, but I figured that these are fairly common functions for testing that would be used when verifying bus traffic. Where would you have them instead?

alevkoy replied:
I figured they should be defined in the .c file with the other bmi160 functions. If you envision using them in tests that would preclude linking in the .c file, I might recommend having some kind of common/utility .c file that could be linked in, but they're OK here too. You might want to leave a comment about where they're expected to be used.

yperess commented on include/zephyr/drivers/emul_sensor.h:
```diff
@@ -26,11 +26,18 @@
 				q31_t *upper, q31_t *epsilon, int8_t *shift);
+	/** Set the offset of a given chanel. */
+	int (*set_offset)(const struct emul *target, enum sensor_channel ch, q31_t *offsets,
+			  int8_t shift);
```
Attributes currently use the old representation of sensor values which are a pain point for many drivers. I purposefully don't want to add more APIs that use `sensor_value` as I believe it will be deprecated in the near future. Reading data will be using the mapping we've introduced with the decoder of channel -> data type. Attributes will likely follow suit. As such, I think it's best to have explicit functions which can provide these better data type bindings to the attributes. It's possible that once we land on the right attribute API, we'll have to collapse these into 1 function, but for now I would prefer not to keep building on the attribute API.

tristan-google replied:
I'm perfectly happy to dump `sensor_value` (we should be using Q31 exclusively in the backend API) or making it a generic pointer to allow for flexibility, but I disagree with creating separate API functions for each kind of attribute. Having a single one (1) reduces code duplication, (2) avoids needing to keep adding more functions when somebody wants the generic test to cover a new attribute, and (3) makes it more awkward for the test to discover supported attributes --with this solution the test would need to test-call each attribute setter function to discover capabilities (and also have a switch block in the test body for fanning out each type of attribute to the appropriate setter)

yperess replied:
Ah, I see. I think this goes beyond just q31_t. I'm getting at the fact that some channel/attribute combinations are bools, some are 1D `q31_t` some are 3D `q31_t` some might be be counters like `uint64_t`. Until we're ready to tackle the attribute API, I don't want to keep building on top of it. Lets assume there's a unified backend API that always accepted `q31_t *values, size_t value_count` as arguments. If an attribute requires a `bool` how would you represent it?

I think it's easier to have separate functions that have the explicit value counts and types as part of the signature. Then when we clean up the attribute API it'll be easy to merge them if/when needed.

yperess replied:
Done

Commit:
bmi160: Fix gyro range map

The range map was sorted wrong since the function
bmi160_range_to_reg_val() that uses it checks for the user value less
than the range component of the bmi160_range struct. This means that
no matter what range is set, the value will always end up 2000dps.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
bmi160: fix value calculation

As an example, when the gyro range is set to 250 deg/sec the scale
was set to 133 where it really should be 133.160058662. This leads
to a 0.12% error in the value returned. By separating the numerator
and denominator, we're able to drastically reduce the error.

Signed-off-by: Yuval Peress <peress@google.com>

teburd reviewed (dismissed):
Seems like lots of C files but with .hpp/.cpp suffixes, not sure if that's what you looking for? I don't think Zephyr has any C++ directly in the tree for things like this does it? My -1 here is for this alone. I don't have a strong opinion on the rest.

More generally though...

Not entirely sure I fully grok the benefit of these tests without HIL to be honest either, were you able to stop/find some legitimate bugs?

Without the physical device around doing a HIL test against it, all the fun quirks and things are likely to fail to pop up. Having to mock out an entire register map to validate that the driver does indeed poke some registers correctly looks like a *lot* of work for questionable benefit to me.

What would it take instead to start having HIL testing with sensors is a question I've been asking myself, given the large number of drivers in the tree, the need to migrate them, and more.

I could very well be alone on this though.

yperess commented on tests/drivers/sensor/bmi160/src/i2c.cpp:
```diff
@@ -0,0 +1,45 @@
+	struct sensor_value value;
+
+	fixture->emul_i2c->bus.i2c->mock_api = &mock_bus_api;
+	mock_bus_api.transfer = [](const struct emul *, struct i2c_msg *msgs, int num_msgs, int) {
+		if (fail_reg_number >= 0 && i2c_is_read_op(&msgs[1]) &&
+		    bmi160_i2c_is_touching_reg(msgs, num_msgs, fail_reg_number)) {
+			return -EIO;
+		}
+		return -ENOSYS;
+	};
```
@teburd this is the reason for C++. I can use a lambda here which makes the test much easier to read since the mock is inside the test function.

teburd replied:
https://docs.zephyrproject.org/latest/develop/languages/cpp/index.html#enabling-c-support

The docs promise C++ support in application code. I don't see any notes about testing code. I generally find C++ incredibly difficult to read but I'm not the sole developer here, maybe writing tests in C++ would be ok, but thats something that likely needs to be brought up in the arch wg. I guess the next question would be what C++ standard? Seemingly C++98 is the only thing guaranteed and I don't believe lambdas are in 98.

yperess replied:
I can't find it in the meeting notes (I guess it wasn't an important comment to write down), but I believe we had this discussion a long time ago in the TSC where we said that tests are fine in C++. This came up when I first added CHRE. I know we don't explicitly ban them:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/9535482/4e2fb6b9-50d1-4dea-abb0-e51f3e75c388)

There are more tests that we can enable via C++ class fixtures later along with the emulators. I personally feel like it gives us a lot of power to test things more easily and quickly.

henrikbrixandersen replied:
I am with @teburd on this one. I do not recall having such a discussion in the TSC and I certainly do not support starting to use C++ in what is otherwise a C project with support for applications written in C++. Sorry.

yperess commented:
> Seems like lots of C files but with .hpp/.cpp suffixes, not sure if that's what you looking for? I don't think Zephyr has any C++ directly in the tree for things like this does it? My -1 here is for this alone. I don't have a strong opinion on the rest.

Added  a comment pointing at why C++. There are more tests coming and having it greatly simplifies the tests. It's even better downstream with Pigweed which I hope to one day have as a module.

> More generally though...
> 
> Not entirely sure I fully grok the benefit of these tests without HIL to be honest either, were you able to stop/find some legitimate bugs?
> 
> Without the physical device around doing a HIL test against it, all the fun quirks and things are likely to fail to pop up. Having to mock out an entire register map to validate that the driver does indeed poke some registers correctly looks like a _lot_ of work for questionable benefit to me.
> 
> What would it take instead to start having HIL testing with sensors is a question I've been asking myself, given the large number of drivers in the tree, the need to migrate them, and more.
> 
> I could very well be alone on this though.

I get what you're saying, but for example all the bugs included in this PR were found through emulation. It's also much more reproducible and runs much faster and cheaper. I can't speak for everyone, but at Google we rely very heavily on these emulators for end to end testing.

henrikbrixandersen commented:
> Without the physical device around doing a HIL test against it, all the fun quirks and things are likely to fail to pop up. Having to mock out an entire register map to validate that the driver does indeed poke some registers correctly looks like a lot of work for questionable benefit to me.
> 
> What would it take instead to start having HIL testing with sensors is a question I've been asking myself, given the large number of drivers in the tree, the need to migrate them, and more.
> 
> I could very well be alone on this though.

I 100% agree. This is close to what was being discussed in the TSC F2F in Prague earlier this year, where I raised this exact concern.

yperess commented:
Removed C++

yperess commented:
@henrikbrixandersen @MaureenHelm any chance to get another review pass before the end of the year? This is a blocker for us to use upstream drivers

MaureenHelm commented on drivers/sensor/bmi160/bmi160.c:
```diff
@@ -251,7 +251,7 @@ struct {
 			      */
 } bmi160_odr_map[] = {
-	{0,    0  }, {0,     780}, {1,     562}, {3,    120}, {6,   250},
+	{0,    0  }, {0,     781}, {1,     562}, {3,    125}, {6,   250},
```
The commit message drops off mid-sentence:
```
There was also some extra conversions from SI->register->SI when
setting the range, this was causing the
```

yperess replied:
Done

MaureenHelm commented on tests/drivers/sensor/accel/src/main.c:
```diff
@@ -55,27 +86,63 @@ static void run_tests_on_accel(const struct device *accel)
 
+	sensor_accel_setup_emulator(fixture.accel_spi, fixture.accel_emul_spi);
+	sensor_accel_setup_emulator(fixture.accel_i2c, fixture.accel_emul_i2c);
+
```
Is this change in the wrong commit? It doesn't seem related to `bmi160: Make changes to align driver to datasheet`

yperess replied:
This is correct, the extra tests here add coverage to the code that was fixed.

yperess commented:
The push above is just a rebase so that the `Compare` button with @MaureenHelm 's requested changes doesn't show the rebase diff. Changes coming in a few minutes.

yperess commented on drivers/sensor/bmi160/emul_bmi160.c:
```diff
@@ -299,6 +287,294 @@ static struct i2c_emul_api bmi160_emul_api_i2c = {
+	case SENSOR_CHAN_ACCEL_Z:
+	case SENSOR_CHAN_ACCEL_XYZ:
+		if (attribute == SENSOR_ATTR_OFFSET) {
+			*min = (q31_t)((int64_t)(-4.857233745 * INT32_MAX) >> 3);
+			*max = (q31_t)((int64_t)(4.857233745 * INT32_MAX) >> 3);
+			*increment = (q31_t)((int64_t)(0.038245935 * INT32_MAX) >> 3);
+			*shift = 3;
+			return 0;
+		}
+		return -EINVAL;
+	case SENSOR_CHAN_GYRO_X:
+	case SENSOR_CHAN_GYRO_Y:
+	case SENSOR_CHAN_GYRO_Z:
+	case SENSOR_CHAN_GYRO_XYZ:
+		if (attribute == SENSOR_ATTR_OFFSET) {
+			*min = (q31_t)(-0.545415391 * INT32_MAX);
+			*max = (q31_t)(0.545415391 * INT32_MAX);
+			*increment = (q31_t)(0.001064651 * INT32_MAX);
```
I expanded them to show how I got them and documented.
